src/state/router.js

```diff
@@ -0,0 +1,142 @@
+
+import { REQUEST_TRIGGER, API_VERSIONS_RECEIVE, REQUEST_SELECT_ENDPOINT } from './actions';
+import { defaultState as defaultRequestState } from './request/reducer';
+import { defaultState as defaultUiState } from './ui/reducer';
+import { loadEndpoints } from './endpoints/actions';
+import { getEndpoints } from './endpoints/selectors';
+
+function getUrlParams() {
+	return new URL( window.location.href ).searchParams;
+}
+
+function encodeString( val ) {
+	return ( val === null || val === undefined ) ? '' : encodeURIComponent( val );
+}
+
+function encodeObject( val ) {
+	return ( typeof val === 'object' ) && Object.keys( val ).length === 0 ? '' : encodeURIComponent( JSON.stringify( val ) );
+}
+
+function decodeString( val ) {
+	return decodeURIComponent( val );
+}
+
+function decodeObject( val ) {
+	return JSON.parse( decodeURIComponent( val ) );
+}
+
+export const getUrlFromState = state => {
+
+	const { request, ui } = state;
+
+	const params = {
+		bodyParams: encodeObject( request.bodyParams ),
+		endpoint: encodeString( request.endpoint ? request.endpoint.pathLabeled : '' ),
+		method: encodeString( request.method ),
+		pathValues: encodeObject( request.pathValues ),
+		queryParams: encodeObject( request.queryParams ),
+		url: encodeString( request.url ),
+		api: encodeString( ui.api ),
+		version: encodeString( ui.version ),
+	};
+
+	// Discard empty params
+	Object.keys( params ).forEach( key => ! params[ key ] && ( delete params[ key ] ) );
+
+	// Construct url
+	const urlParams = new URLSearchParams( params );
+	const url = window.location.origin + window.location.pathname + '?' + urlParams.toString();
```

Comment:
I think we can create a `new URL( window.location )` here and then directly use our search params to replace any existing ones.

```js
const url = new URL( window.location );
url.search = urlParams;

return url;
```